agent/agent/process/size_limits.py

```diff
@@ -0,0 +1,120 @@
+            if Hold.Type.SOURCE_OVERSIZE in trigger.after.hold_types:
+                return
+
+            emit(AddHold(creator=self.agent,
+                         hold_type=Hold.Type.SOURCE_OVERSIZE,
+                         hold_reason=msg))
+
+        # If the submission is on hold due to oversize, and the submission is
+        # no longer oversize, remove the hold.
+        else:
+            for event_id, hold in trigger.after.holds.items():
+                if hold.hold_type is Hold.Type.SOURCE_OVERSIZE:
+                    emit(RemoveHold(creator=self.agent, hold_event_id=event_id,
+                                    hold_type=Hold.Type.SOURCE_OVERSIZE,
+                                    removal_reason=msg))
+
+
+class CheckPDFSize(Process):
+    """When a PDF is compiled, check for oversize."""
+
+    def handle_compiler_exception(self, exc: Exception) -> None:
+        """Handle exceptions raised when calling the compiler service."""
+        exc_type = type(exc)
+
+        if exc_type in (exceptions.BadResponse, exceptions.ConnectionFailed):
+            raise Recoverable('Encountered %s; try again' % exc) from exc
+        elif exc_type is exceptions.RequestFailed:
+            if exc.status_code >= 500:
+                msg = 'Compiler service choked: %i' % exc.status_code
+                raise Recoverable(msg) from exc
+            self.fail(exc, 'Unrecoverable exception: %i' % exc.status_code)
+        self.fail(exc, 'Unhandled exception')
+
+    @step(max_retries=None)
+    def get_size(self, previous: Optional, trigger: Trigger,
+                 emit: Callable) -> int:
+        """Get the size of the compilation from the compiler service."""
+        try:
+            source_id = trigger.after.source_content.identifier
+            source_state = trigger.after.source_content.checksum
+        except AttributeError as exc:
+            self.fail(exc, message='Missing compilation or post-event state')
+        compilation_id = Compilation.get_identifier(source_id, source_state)
+        scopes = get_compiler_scopes(compilation_id)
+        token = get_system_token(__name__, self.agent, scopes)
+
+        try:
+            stat = compiler.Compiler.get_status(source_id, source_state, token,
+                                                Compilation.Format.PDF)
+        except Exception as exc:
+            self.handle_compiler_exception(exc)
+        if stat.status is Compilation.Status.IN_PROGRESS:
+            raise Recoverable('Compilation is stil in progress; try again')
+        elif stat.status is Compilation.Status.FAILED:
+            self.fail(message='Compilation failed; cannot get size of PDF')
+        return stat.size_bytes
+
+    @step()
+    def evaluate_size(self, size_bytes: int, trigger: Trigger,
+                      emit: Callable) -> int:
+        """Add or remove holds as appropriate."""
+        msg = "PDF is %i bytes" % size_bytes
+        if size_bytes > trigger.params['PDF_LIMIT']:
```

Comment:
It might be helpful to commemorate the `PDF_LIMIT` in the message that gets logged as the hold reason.

Reply:
Changed in 043e5b2:

```python
    @step()
    def evaluate_size(self, size_bytes: int, trigger: Trigger,
                      emit: Callable) -> int:
        """Add or remove holds as appropriate."""
        pdf_limit_bytes = trigger.params['PDF_LIMIT_BYTES']
        if size_bytes > pdf_limit_bytes:
            if Hold.Type.PDF_OVERSIZE in trigger.after.hold_types:
                return      # Already on hold for this reason; nothing to do.
            emit(AddHold(
                creator=self.agent,
                hold_type=Hold.Type.PDF_OVERSIZE,
                hold_reason=f'PDF is {size_bytes} bytes, exceeds limit of'
                            f' {pdf_limit_bytes} bytes'))
        else:
            # If the submission is on hold due to oversize, remove the hold.
            for event_id, hold in trigger.after.holds.items():
                if hold.hold_type is Hold.Type.PDF_OVERSIZE:
                    emit(RemoveHold(
                        creator=self.agent,
                        hold_event_id=event_id,
                        hold_type=Hold.Type.PDF_OVERSIZE,
                        removal_reason=f'PDF is {size_bytes} bytes, below'
                                       f' limit of {pdf_limit_bytes} bytes'))
```

Reply:
Thanks, looks good.